OpenRA.Mods.Common/Widgets/ObserverArmyIconsWidget .cs

```diff
@@ -0,0 +1,220 @@
+		public int IconSpacing = 1;
+
+		Dictionary<string, Tuple<ActorInfo, int, int>> unitLookup;
+		float2 iconSize;
+		public int MinWidth = 240;
+
+		public ActorInfo TooltipUnit { get; private set; }
+		public Func<ActorInfo> GetTooltipUnit;
+
+		public readonly string TooltipTemplate = "ARMY_TOOLTIP";
+		public readonly string TooltipContainer;
+
+		readonly Lazy<TooltipContainerWidget> tooltipContainer;
+		readonly List<ActorInfo> armyUnits = new List<ActorInfo>();
+		readonly List<Rectangle> armyIconsBounds = new List<Rectangle>();
+
+		int lastIconIdx;
+
+		[ObjectCreator.UseCtor]
+		public ObserverArmyIconsWidget(World world, WorldRenderer worldRenderer)
+		{
+			this.world = world;
+			this.worldRenderer = worldRenderer;
+
+			GetTooltipUnit = () => TooltipUnit;
+			tooltipContainer = Exts.Lazy(() =>
+				Ui.Root.Get<TooltipContainerWidget>(TooltipContainer));
+
+			Init();
+		}
+
+		protected ObserverArmyIconsWidget(ObserverArmyIconsWidget other)
+			: base(other)
+		{
+			GetPlayer = other.GetPlayer;
+			world = other.world;
+			worldRenderer = other.worldRenderer;
+
+			IconWidth = other.IconWidth;
+			IconHeight = other.IconHeight;
+			IconSpacing = other.IconSpacing;
+			iconSize = new float2(IconWidth, IconHeight);
+
+			MinWidth = other.MinWidth;
+
+			TooltipUnit = other.TooltipUnit;
+			GetTooltipUnit = () => TooltipUnit;
+
+			TooltipTemplate = other.TooltipTemplate;
+			TooltipContainer = other.TooltipContainer;
+
+			tooltipContainer = Exts.Lazy(() =>
+				Ui.Root.Get<TooltipContainerWidget>(TooltipContainer));
+
+			Init();
+		}
+
+		void Init()
+		{
+			var buildable = world.Map.Rules.Actors.Values
+				.Where(x =>
+					x.Name[0] != '^' &&
+					x.HasTraitInfo<BuildableInfo>())
+				.SelectMany(x => x.TraitInfo<BuildableInfo>().Queue, (ai, q) => new Tuple<string, ActorInfo, BuildableInfo>(q, ai, ai.TraitInfo<BuildableInfo>()));
+
+			var pqueue = world.Map.Rules.Actors.Values
+				.SelectMany(a => a.TraitInfos<ProductionQueueInfo>());
+
+			unitLookup = buildable.Join(pqueue, tuple => tuple.Item1, pq => pq.Type, (tuple, info) => new { tuple.Item2.Name, Ai = tuple.Item2, tuple.Item3.BuildPaletteOrder, info.DisplayOrder })
+				.Distinct()
+				.ToDictionary(t => t.Name, o => new Tuple<ActorInfo, int, int>(o.Ai, o.DisplayOrder, o.BuildPaletteOrder));
+		}
+
+		public override void Draw()
+		{
+			armyUnits.Clear();
+			armyIconsBounds.Clear();
+
+			var player = GetPlayer();
+			if (player == null)
+				return;
+
+			var playerStatistics = player.PlayerActor.TraitOrDefault<PlayerStatistics>();
+
+			var items = playerStatistics.Units
+				.Where(i => i.Value.Count > 0)
+				.OrderBy(i => unitLookup[i.Key].Item2)
+				.ThenBy(i => unitLookup[i.Key].Item3)
+				.ToList();
+
+			Bounds.Width = items.Count * (IconWidth + IconSpacing);
+
+			var queueCol = 0;
+			foreach (var unit in items)
+			{
+				var name = unit.Key;
+				var actorInfo = unitLookup[name].Item1;
+
+				var rsi = actorInfo.TraitInfo<RenderSpritesInfo>();
```

Comment:
We could save a bunch of lookups here by defining a class that holds the icon, BuildableInfo, and palette, and then keeping a `Cache<string, ActorData>`